warehouse/forklift/legacy.py

```diff
@@ -807,24 +807,27 @@ def file_upload(request):
             ),
         )
 
-    # Ensure that user has a verified, primary email address. This should both
-    # reduce the ease of spam account creation and activity, as well as act as
-    # a forcing function for https://github.com/pypa/warehouse/issues/3632.
-    # TODO: Once https://github.com/pypa/warehouse/issues/3632 has been solved,
-    #       we might consider a different condition, possibly looking at
-    #       User.is_active instead.
-    if not (request.user.primary_email and request.user.primary_email.verified):
-        raise _exc_with_message(
-            HTTPBadRequest,
-            (
-                "User {!r} does not have a verified primary email address. "
-                "Please add a verified primary email before attempting to "
-                "upload to PyPI. See {project_help} for more information."
-            ).format(
-                request.user.username,
-                project_help=request.help_url(_anchor="verified-email"),
-            ),
-        ) from None
+    # These checks only make sense when our authenticated identity is a user,
+    # not a project identity (like OIDC-minted tokens.)
+    if request.user:
+        # Ensure that user has a verified, primary email address. This should both
+        # reduce the ease of spam account creation and activity, as well as act as
+        # a forcing function for https://github.com/pypa/warehouse/issues/3632.
+        # TODO: Once https://github.com/pypa/warehouse/issues/3632 has been solved,
+        #       we might consider a different condition, possibly looking at
+        #       User.is_active instead.
+        if not (request.user.primary_email and request.user.primary_email.verified):
+            raise _exc_with_message(
+                HTTPBadRequest,
+                (
+                    "User {!r} does not have a verified primary email address. "
+                    "Please add a verified primary email before attempting to "
+                    "upload to PyPI. See {project_help} for more information."
+                ).format(
+                    request.user.username,
+                    project_help=request.help_url(_anchor="verified-email"),
+                ),
+            ) from None
```

Comment:
Flagging for review: the entire upload endpoint assumes that the request's identity is a user. I've handled most of these by checking whether `request.user is None` (with the precondition that `request.identity is not None`).